src/dblib_stubs.c

```diff
@@ -136,7 +178,10 @@ static int err_handler(DBPROCESS *dbproc, int severity, /* in syberror.h */
   /* FIXME: allow the handler to return whether to continue trying on
      timeouts?  Separate handler for timeouts?? */
 
-  CAMLreturn(INT_CANCEL); /* should not return */
+  if (had_lock) {
+    release_lock(dbproc);
+  }
+  CAMLreturn(INT_CANCEL);
```

Comment:
Another case where this can return if the error handle doesn't throw an exception.